### PR TITLE
[Concurrency] Handle partial application of actor-isolated methods.

### PR DESCRIPTION
**Pull request by DougGregor**
Implement appropriate isolation checking for partial application of actor-isolated methods.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 1]
```diff
@@ -338,6 +338,69 @@ class IsolationRestriction {
...
+  };
+}
+
+/// Try to decompose a call that might be an invocation of a partial apply
+/// thunk.
+static Optional<PartialApplyThunkInfo> decomposePartialApplyThunk(
+    ApplyExpr *apply, Expr *parent) {
+  // Check for a call to the outer closure in the thunk.
+  auto outerAutoclosure = dyn_cast<AutoClosureExpr>(apply->getFn());
+  if (!outerAutoclosure ||
```
Comment (theblixguy) [inline lib/Sema/TypeCheckConcurrency.cpp:356]: I think you can use `getUnwrappedCurryThunkExpr()` on `outerAutoclosure` to dig out the `DeclRefExpr` of the function.

Comment (DougGregor) [reply lib/Sema/TypeCheckConcurrency.cpp:356]: I didn't know this function existed! I'll figure out how to use it, thank you.

**Comment by slavapestov** [comment]
Might be worth adding some SILGen tests for the valid cases.

**Comment by DougGregor** [comment]
> Might be worth adding some SILGen tests for the valid cases.

We don't have any SILGen support for asynchronous functions yet, so we can't write these tests meaningfully.

File: lib/Sema/TypeCheckConcurrency.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/TypeCheckConcurrency.cpp:356]: Heh, good news. Doing do improved test coverage for `getUnwrappedCurryThunkExpr()`, so now both my new client and all existing clients handle more curry think ASTs properly.

**Comment by DougGregor** [comment]
Windows issue is unrelated:
```
mt.exe : general error c101008d: Failed to write the updated manifest to the resource of file "dispatch_timer_short.exe". Access is denied.
```

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
Source compatibility suite testing is still being adjusted for the master -> main transition.

**Pull request merged**